Adding details to types in PHPDoc

orklah opened the pull request:
This PR propose detailing phpdoc for the most part. Most types were generated directly by Psalm and refined by hand.

orklah committed:
psalm fixes

greg0ire commented on lib/Doctrine/ORM/AbstractQuery.php:
```diff
@@ -323,7 +325,7 @@ public function getParameters()
      * @param mixed $key The key (index or name) of the bound parameter.
      *
-     * @return Query\Parameter|null The value of the bound parameter, or NULL if not available.
+     * @return Parameter|false|null The value of the bound parameter, or NULL if not available.
```
I don't think `false` can actually be returned here: if `$filteredParmeters` is empty, we return `null`

orklah replied:
You're right. I guess we could make Psalm understand this somehow to solve this on future levels

greg0ire commented on lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php:
```diff
@@ -151,9 +152,8 @@ public function iterate($stmt, $resultSetMapping, array $hints = [])
-     * @return iterable<mixed>
-     *
      * @psalm-param array<string, mixed> $hints
+     * @psalm-return Generator<int, (array|mixed), mixed, void>
```
What does this syntax mean? I'd expect no more than one comma inside the first level of `<>`

orklah replied:
Generator is kinda special and Psalm support up to 4 templated types for it: https://psalm.dev/docs/annotating_code/type_syntax/object_types/
Strangely, Psalm's doc says that the last two are mixed by default, but Psalter generated a void in 4th position instead of mixed. It doesn't seem to provoke errors up to Psalm lvl 4 one way or another, I removed the last two parameters

greg0ire replied:
Oh right, how did I forget about this 👍 Anyway here it doesn't matter indeed, removing them is good.

greg0ire commented on lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php:
```diff
@@ -151,9 +152,8 @@ public function iterate($stmt, $resultSetMapping, array $hints = [])
-     * @return iterable<mixed>
-     *
      * @psalm-param array<string, mixed> $hints
+     * @psalm-return Generator<int, (array|mixed)>
```
:thinking: the union of `array` and `mixed` is `mixed`, so what's the point?

orklah replied:
fixed

greg0ire commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -1138,12 +1139,11 @@ public function enableAssociationCache($fieldName, array $cache)
-     * @param array  $cache
      *
-     * @return mixed[]
+     * @return (int|string)[]
```
Are parenthesis understood by IDEs?

orklah replied:
Understood by VS Code but not by phpstorm. Fixed

greg0ire commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -1409,7 +1425,9 @@ public function getSqlResultSetMapping($name)
      * Gets all sql result set mappings of the class.
      *
-     * @psalm-return array<string, array<string, mixed>>
+     * @return (array|string)[][]
```
:thinking: phpcs might be happy with this for now, but maybe this should become `(mixed[]|string)[][]` or something more precise to be future proof

orklah replied:
fixed

greg0ire commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -1869,7 +1911,8 @@ protected function _validateAndCompleteOneToManyMapping(array $mapping)
      * @psalm-param array<string, mixed> $mapping The mapping to validate & complete.
-     * @psalm-return array{isOwningSide: mixed, orphanRemoval: bool}
+     *
+     * @psalm-return array{mappedBy: (mixed|null), inversedBy: (mixed|null), isOwningSide: bool, sourceEntity: class-string, targetEntity: string, fieldName: mixed, fetch: mixed, cascade: array<string>, isCascadeRemove: bool, isCascadePersist: bool, isCascadeRefresh: bool, isCascadeMerge: bool, isCascadeDetach: bool, type: int, originalField: string, originalClass: class-string, joinTable?: (Arra ...
```
Please use linebreaks

orklah replied:
fixed

greg0ire commented on lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:
```diff
@@ -821,9 +823,9 @@ private function columnToArray(SimpleXMLElement $fieldMapping)
      * Parse / Normalize the cache configuration
      *
-     * @return mixed[]
+     * @return (mixed|string|null)[]
```
```suggestion
     * @return mixed[]
```

orklah replied:
fixed

greg0ire commented on lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php:
```diff
@@ -399,7 +399,7 @@ protected function getDeleteSQL(PersistentCollection $collection)
      * Internal note: Order of the parameters must be the same as the order of the columns in getDeleteSql.
      *
-     * @return mixed[]
+     * @return list<mixed>
```
Should this be duplicated because of the `list` part?

orklah replied:
list is now an alias of array in phpstorm. Not sure in VS Code

greg0ire commented on lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:
```diff
@@ -1856,7 +1856,7 @@ public function expandParameters($criteria)
      * @return mixed[][]
      *
-     * @psalm-return array{0: array, 1: list<mixed>}
+     * @psalm-return array{0: array, 1: list<(int|string|null)>}
```
The parenthesis don't look necessary

orklah replied:
fixed

greg0ire commented on lib/Doctrine/ORM/Persisters/SqlValueVisitor.php:
```diff
@@ -84,7 +84,7 @@ public function walkValue(Value $value)
      * @return mixed[][]
      *
-     * @psalm-return array{0: array, 1: array}
+     * @psalm-return array{0: array, 1: array<array>}
```
Let's use `mixed[]` here and there?

orklah replied:
replaced by array<mixed> to keep coherence

greg0ire commented on lib/Doctrine/ORM/Query/QueryExpressionVisitor.php:
```diff
@@ -51,7 +50,7 @@ class QueryExpressionVisitor extends ExpressionVisitor
     private $expr;
 
-    /** @var mixed[] */
+    /** @var array<int, mixed> */
```
Maybe we should use `list` here and in the constructor?

orklah replied:
replaced here, the property was not from constructor though

greg0ire commented on lib/Doctrine/ORM/Query/ResultSetMapping.php:
```diff
@@ -428,7 +430,7 @@ public function isScalarResult($columnName)
      * @param string $alias
      *
-     * @return string
+     * @return class-string
```
This is not understood by IDEs is it?

orklah replied:
It's now an alias of string in PHPStorm. Not sure in VS Code

greg0ire commented:
How closer to Psalm and PHPStan's next levels does this bring us?

greg0ire commented:
@vladyslavstartsev @simonberger please review :pray:

orklah commented:
> How closer to Psalm and PHPStan's next levels does this bring us?

It wasn't quite my metric, but with issues up to level 3, it fixes 63 issues, it revealed 14 new ones and upped the type coverage from 90.4980% to 90.5546%

greg0ire commented:
Why are there 2 commits? What is the difference between "psalm fixes" and just "fixes"?

orklah commented:
!'m still working on that, I'll squash them later

orklah commented:
This should be good now :)

This has been rebased and I fixed the new issues related to the baselines that were introduced.

I had to change a few things:
- I updated phpstan version. The version that was used before didn't support PHP8 so I couldn't work with it in my dev environment
- I introduced phpstan/phpstan-webmozart-assert. The library is used at some point and phpstan needs the plugin to be able to understand what it does
- I added a set php version to phpstan config. This will ensure everybody runs phpstan against the same version (namely the oldest version supported by the branch). This is needed to have a coherent baseline generated no matter what php version used to run phpstan itself (I was getting issues with different errors in PHP8 than in PHP7.4 on CI and even in PHP 7.2). Note that this is not the recommandation for this: https://github.com/phpstan/phpstan/issues/4871#issuecomment-821778532 but going further is not in the scope of this PR at all. It may be adressed later.

greg0ire commented:
Your last message has bullet points containing interesting information. Could you please make one commit per bullet point, and reuse what you wrote to create a commit messages? Some of these commits should probably target 2.8.x. For instance, to avoid issues when merging up, it would be great to have the same versions of phpstan on all branches.

The second one can stay on the 2.9.x PR I think, because it will have no effect on the CI of the end user.

The third one should be contributed to 2.8.x, since it does not affect the code itself but rather the build process, and as you pointed out, it seems important to have the same messages, especially when merging up (we don't want to have complicated things to do when merging up).

And finally, some changes in this PR are type improvements, while some others are type fixes (the types were just plain wrong before, for instance when we use `int` instead of `int|false` as a return type. Using `git reset origin/2.9.x && git add -p` please separate the changes in 2 sets, then cherry-pick one of those on a new branch created from 2.8.x, then we will merge it, and merge the result up in 2.9.x

greg0ire reviewed (approved):
I learnt today that 2.9.0 was going to be released really soon, so there is no need to backport the "fixes" to 2.8.x

greg0ire commented:
Thanks @orklah !
